### ISSUE TITLE
Page Collections

### ISSUE THREAD

**Issue by bokand** [WICG/proposals#92]
## Introduction

Today, linking on the web is largely one-to-one. A user wishing to share a collection of resources must share the links individually and recipients must click through each link individually and, if their UI supports it, manually group the opened resources.

This is a proposal to enable users and authors to collect multiple resources into a single link. An agent navigating to such a link is able to load and present all the link's resources in a grouping UI. We're calling such a link a "page collection", e.g.:

```
uri-list:https://example.com/pageA;https://en.wikipedia.org/wiki/URL;https://w3.org
```

Page collections aims to allow the user agent to be more helpful in opening and presenting multiple related pages, for example, by opening each page as a tab in a tab group.

No grouping UI is suggested or mandated by this proposal; how such a link is presented will be left to the discretion of the user agent.

### Summary of proposed web-facing changes

* (loosely) Specify web browser handling of `text/uri-list` media types.
* Introduce a `uri-list:` URI scheme to allow users to easily share collection links.
* Add an opt-in for for anchors (i.e. `<a>`) to non-uri-list schemes to allow opening a collection (i.e. a regular `http(s)` link that responds with `Content-Type: text/uri-list`.
* Define a permission model for when a collection link may open multiple windows.
* Use content negotiation to allow servers to feature detect user agents capable of handling `text/uri-list` responses.

[Read the complete Explainer](https://github.com/bokand/page-collection). 

## Feedback (Choose One)

I welcome feedback in this thread, but encourage you to file bugs against [the Explainer](https://github.com/bokand/page-collection).

[explainer]: https://github.com/bokand/page-collection "Page Collections"

**Comment by raphaellouisandress** [comment]
Hi @bokand, how are you?

A curious question: would it be possible to include text fragments in uri-list?
```python
uri-list:ws://linux-gamer.com/2023/docker-linux/#:~:text=25,567;ftp://linux-gamer.com/2023/docker-linux/#:~:text=25,567;https://w3.org
uri-list:UTF8,25,567,ws://linux-gamer.com/2023/docker-linux/hashID;UTF8,25,567,ftp://linux-gamer.com/2023/docker-linux/hashID
```

**Comment by bokand** [comment]
Yup, text fragments are just URL fragments so there's no reason that shouldn't work. Implementors would have to exercise a bit of caution, for [reasons listed](https://github.com/bokand/page-collection#:~:text=Leaking%20data,-Pages%20authored%20in) in the Abuse and Security section, but I expect those wouldn't be a blocker. Indeed, the scroll-to-text spec already [prevents scrolling in background tabs](https://wicg.github.io/scroll-to-text-fragment/#:~:text=no%20scrolling%20is%20performed%20while%20the%20document%20is%20in%20the%20background).

**Comment by raphaellouisandress** [comment]
> Yup, text fragments are just URL fragments so there's no reason that shouldn't work.

Could you include this idea in your proposal?

**Comment by AdamSobieski** [comment]
@bokand, brainstorming and expanding upon the [proposed syntax ideas](https://github.com/bokand/page-collection), one could consider an HTML5 [custom element](https://html.spec.whatwg.org/dev/custom-elements.html), let us call it: `<anchor-list>`, with which to express a [navigation](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigate) with respect to multiple [navigables](https://html.spec.whatwg.org/multipage/document-sequences.html#navigable).
```xml
<anchor-list>
  <content>hypertext</content>
  <anchors>
    <a target="frame1" href="file1.html#loc-1" />
    <a target="frame2" href="file2.html#loc-2" />
  </anchors>
</anchor-list>
```

Then, the matter is one of exploring how to express those semantics using a single `<a>` element, how to encode those semantics into a URI string.

Adding a `?` after the proposed `multi:` scheme, as inspired by [the Magnet URI scheme](https://en.wikipedia.org/wiki/Magnet_URI_scheme), and considering [the expression of arrays in query strings](https://stackoverflow.com/questions/6243051/how-to-pass-an-array-within-a-query-string), one could express a multilink in a manner resembling:
```html
<a href="multi:?target[]=frame1&href[]=file1.html#loc-1&target[]=frame2&href[]=file2.html#loc-2">hypertext</a>
```

This syntax also supports the naming of tab groups for page collections:
```html
<a href="multi:?target[]=_blank&href[]=a.html&target[]=_blank&href[]=b.html&group-name=Some%20Tabs">hypertext</a>
```

If `_blank` was the default `target` value for multilinks, the previous example could be more succinctly expressed:
```html
<a href="multi:?href[]=a.html&href[]=b.html&group-name=Some%20Tabs">hypertext</a>
```

In addition to the `[]` notation for expressing arrays in query strings, one could utilize a dot-delimited notation:
```html
<a href="multi:?target.0=_blank&href.0=a.html&target.1=_blank&href.1=b.html&group-name=Some%20Tabs">hypertext</a>
```
```html
<a href="multi:?href.0=a.html&href.1=b.html&group-name=Some%20Tabs">hypertext</a>
```

Thank you. What do you think of these ideas?

**Comment by AdamSobieski** [comment]
Also, instead of `text/uri-list` resources, what about creating a new text-, XML-, or JSON-based format with a new MIME type?

A more expressive format could be of use for [agent-driven content negotiation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/Content_negotiation#agent-driven_negotiation) and with the [`data:` URI scheme](https://en.wikipedia.org/wiki/Data_URI_scheme).

A new format could enable the following expressiveness:

```xml
<nav>
  <seq name="Page Collection 1">
    <a href="baz1.html" />
    <a href="baz2.html" />
  </seq>
  <seq name="Page Collection 2">
    <a href="baz3.htm" />
    <a href="baz4.htm" />
  </seq>
  <seq name="Page Collection 3">
    <a href="baz5.html" />
    <alt>
      <a hreflang="en" type="application/epub+zip" media="(scripting: enabled)" href="baz6-js-en.epub" />
      <a hreflang="fr" type="application/epub+zip" media="(scripting: enabled)" href="baz6-js-fr.epub" />
      <a hreflang="en" type="application/epub+zip" href="baz6-en.epub" />
      <a hreflang="fr" type="application/epub+zip" href="baz6-fr.epub" />
      <a hreflang="en" type="application/xhtml+xml" href="baz6-en.html" />
      <a hreflang="fr" type="application/xhtml+xml" href="baz6-fr.html" />
    </alt>
  </seq>
</nav>
```

The above example would navigate to six destinations, each opening in its own new tab (`_blank` target being default). These tabs would be organized into three groups. The sixth resource would be determined per client-side agent-driven content negotiation.
